Fix generic method resolution in composite R2R images

trylek opened the pull request:
According to my recent investigation, worse perf in the presence
of composite R2R images is due to the inability to resolve certain
generic methods in them; as an initial improvement in that direction
I propose changing generic instance signature matching to
always load assemblies referenced in their signatures. I'm locally
seeing this fix cancel the previous composite-specific regression
I originally measured on the webapi project. Please let me know
if you have different suggestions how to tackle this issue.

Thanks

Tomas

/cc: @dotnet/crossgen-contrib 
Fixes: #64170

trylek committed:
Fix generic method resolution in composite R2R images

According to my recent investigation, worse perf in the presence
of composite R2R images is due to the inability to resolve certain
generic methods in them; as an initial improvement in that direction
I propose changing generic instance signature matching to
always load assemblies referenced in their signatures. I'm locally
seeing this fix cancel the previous composite-specific regression
I originally measured on the webapi project. Please let me know
if you have different suggestions how to tackle this issue.

Thanks

Tomas

trylek commented:
/azp run runtime-coreclr crossgen2 outerloop

jkotas commented on src/coreclr/vm/zapsig.cpp:
```diff
@@ -343,7 +340,7 @@ BOOL ZapSig::GetSignatureForTypeHandle(TypeHandle      handle,
             DWORD ix = CorSigUncompressData(pSig);
             CONTRACT_VIOLATION(ThrowsViolation|GCViolation);
-            pModule = pZapSigContext->GetZapSigModule()->GetModuleFromIndexIfLoaded(ix);
+            pModule = pZapSigContext->GetZapSigModule()->GetModuleFromIndex(ix);
```
Can this cause random assemblies to get loaded early?

trylek replied:
Well, with a probability corresponding to hash collisions (theoretically 2^-32) it can load more assemblies than required. One possible way to mitigate this would be to add a third method, something like <code>GetModuleFromIndexIfLoadedOrComposite</code>, that would only apply the logic of loading extra assemblies in the presence of a composite image thus removing any impact on non-composite code (even though in practice I doubt this happens very often as single-assembly crossgen files shouldn't precompile generic instantiations involving references to other modules. Without some form of loading the reference assembly I currently don't see how to make the precompiled generic instantiations in composite images useful.

jkotas replied:
> Well, with a probability corresponding to hash collisions (theoretically 2^-32)

I believe that the probability can be much higher in aggregate due birthday paradox.

>  Without some form of loading the reference assembly I currently don't see how to make the precompiled generic instantiations in composite images useful.

I think you just need to resolve the reference assembly from the already loaded assemblies to see whether it matches. All constituents of the generic method that we are trying to look up are loaded already. It means that we are not going to miss anything when we do not load any new assemblies.

trylek replied:
Fixed in 2nd commit, thanks for pointing that out.

trylek committed:
Fix assembly caching in the presence of R2R composite images

Investigating R2R perf deficiencies in the presence of composite
images I found out that many generic methods don't get resolved
from the composite image. Ultimately I believe the problem was
inconsistent caching of the component assemblies. In particular,
DecodeModuleFromIndex uses the native metadata assembly ref cache
to resolve component assemblies but DecodeModuleFromIndexIfLoaded
doesn't consult it and direct load of an arbitrary assembly that
is a component to the composite image also doesn't set the cache
entry. This change fixes these deficiencies.

Thanks

Tomas

trylek commented:
Based on Jan Kotas' PR feedback I drilled deeper into the issue and I believe I have finally nailed down the problem. Turns out the crucial method <code>DecodeModuleFromIndexIfLoaded</code> fails to resolve the offending assembly (System.Linq in my original investigation) due to several interesting reasons:

1) While <code>DecodeModuleFromIndex</code> uses <code>GetNativeMetadataAssemblyRefFromCache</code> / <code>SetNativeMetadataAssemblyRefInCache</code> to cache the assembly resolution, <code>DecodeModuleFromIndexIfLoaded</code> never consulted the same cache. This PR fixes that deficiency.

2) Even though <code>DecodeModuleFromIndex</code> updates the assembly cache, it doesn't kick in when the assembly is loaded directly, not as a referenced entry point. I have fixed <code>NativeImage::Open</code> to register the component assembly.

3) There seems to be an additional assembly cache tied to the assembly binder that should resolve the assembly as a backdrop but it doesn't. I haven't yet investigated this secondary issue and I'm not trying to fix it in this PR.

Thanks

Tomas

jkotas commented:
Yes, this approach looks better. I see number of test failures in the crossgen2 legs. Are any of them related to this change?

trylek commented:
I believe not, I see some of them in runs before my change, I'm working on trying to repro them locally and I'll rerun the outerloop after merging in the new addition to issues.targets.

trylek commented:
So there are basically three categories of the CG2 outerloop failures now: 

1) Crossgen2 compilation-time crashes in tests recently added by @jkoritzinsky under Interop\DisabledRuntimeMarshalling, e.g. here: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-64247-head-7288d3d7dce34fca97/Interop/1/console.7e0fc92f.log?sv=2019-07-07&se=2022-02-16T01%3A11%3A42Z&sr=c&sp=rl&sig=YeXcpRNNiiAogWSRaluJ1v6SF9REZeBz5LpiS8uR42I%3D. I'm debugging the problem now.

2) Field offset mismatch on 64-bit platforms in the Runtime_60035 regression test, e.g. here: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-64247-head-7288d3d7dce34fca97/JIT.Regression/1/console.70a2b72d.log?sv=2019-07-07&se=2022-02-16T01%3A11%3A42Z&sr=c&sp=rl&sig=YeXcpRNNiiAogWSRaluJ1v6SF9REZeBz5LpiS8uR42I%3D. I'm not yet sure what to do about this one, there's a possibility it's a fallout of Jeremy's type layout refactoring but the CG2 outerloop runs only receive random attention so I'm not 100% sure.

3) A bunch of one-off random crashes, some of them on arm64, mostly in various threading tests. I've ignored them for now.

jkoritzinsky commented:
The crashes in item 1 should have been fixed by https://github.com/dotnet/runtime/pull/64286

trylek commented:
Wonderful, thanks Jeremy for the immediate response - I've just unsuccessfully tried to repro it locally so that explains it I guess. In light of this fact I'll take a look at item 2 and then I'll run a new CG2 outerloop run.

trylek commented:
/azp run runtime-coreclr outerloop

trylek commented:
/azp run runtime-coreclr outerloop

trylek commented:
I believe I have addressed all review feedback so far and the only remaining failures in the runs are the known timeouts in the Vector tests; from my viewpoint the change is ready to be merged in, please let me know if you're aware of any remaining deficiencies that need fixing first.
